presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -700,6 +705,90 @@ private void validateMaterialziedViewQueryPlan(Statement query)
+
+            Scope viewScope = process(node.getTarget(), scope);
+            Map<SchemaTableName, Expression> tablePredicates = extractTablePredicates(viewName, node.getWhere(), viewScope, metadata, session);
+
+            Query viewQuery = parseView(view.getOriginalSql(), viewName, node);
+            Query refreshQuery = tablePredicates.containsKey(toSchemaTableName(viewName)) ?
+                    buildQueryWithPredicate(viewQuery, tablePredicates.get(toSchemaTableName(viewName)))
+                    : viewQuery;
+            process(refreshQuery, scope);
+
+            TableHandle tableHandle = metadata.getTableHandle(session, viewName)
+                    .orElseThrow(() -> new SemanticException(MISSING_MATERIALIZED_VIEW, node, "Materialized view '%s' does not exist", viewName));
+            Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+            List<ColumnHandle> targetColumnHandles = metadata.getTableMetadata(session, tableHandle).getColumns().stream()
+                    .filter(column -> !column.isHidden())
+                    .map(column -> columnHandles.get(column.getName()))
+                    .collect(toImmutableList());
+            analysis.setRefreshMaterializedViewAnalysis(new Analysis.RefreshMaterializedViewAnalysis(
+                    tableHandle,
+                    targetColumnHandles,
+                    refreshQuery));
+
+            return createAndAssignScope(node, scope, Field.newUnqualified("rows", BIGINT));
+        }
+
+        private Optional<RelationType> analyzeBaseTableForRefreshMaterializedView(
+                Table baseTable,
+                Optional<Scope> scope)
+        {
+            checkState(analysis.getStatement() instanceof RefreshMaterializedView, "Not analyzing RefreshMaterializedView statement");
+
+            RefreshMaterializedView refreshMaterializedView = (RefreshMaterializedView) analysis.getStatement();
+            QualifiedObjectName viewName = createQualifiedObjectName(session, refreshMaterializedView.getTarget(), refreshMaterializedView.getTarget().getName());
+
+            // Analyzer will check for SELECT permission on materialized view table, but REFERSH MATERIALIZED VIEW only needs INSERT permission.
+            // So use a new StatementAnalyzer instance with access check disabled to analyze the materialized view table.
+            StatementAnalyzer analyzer = new StatementAnalyzer(
+                    analysis,
+                    metadata,
+                    sqlParser,
+                    new AllowAllAccessControl(),
+                    session,
+                    warningCollector);
+            Scope viewScope = analyzer.analyze(refreshMaterializedView.getTarget(), Scope.create());
+
+            Map<SchemaTableName, Expression> tablePredicates = extractTablePredicates(viewName, refreshMaterializedView.getWhere(), viewScope, metadata, session);
```

Comment:
Do we want cache this in the Analysis object so that we dont need to process it two times?

Reply:
I think it's fine to compute it on the fly since it's not expensive. Thought once about caching it but the cache key will be tricky, which is view_name + predicate.